netket/optimizer/qgt/qgt_jacobian_dense.py

```diff
@@ -127,8 +147,87 @@ def to_dense(self) -> jnp.ndarray:
             diag = jnp.diag(self.scale**2)
 
         # concatenate samples with real/Imaginary dimension
-        O = O.reshape(-1, O.shape[-1])
-        return mpi.mpi_sum_jax(O.conj().T @ O)[0] + self.diag_shift * diag
+        if self.mode == "imag":
+            # Equivalent to Jr.T@Ji - Ji.T@Jr
+            flip_sign = jnp.array([1, -1]).reshape(1, 2, 1)
+            Ol = (flip_sign * O).reshape(-1, O.shape[-1])
+            Or = jnp.flip(O, axis=1).reshape(-1, O.shape[-1])
+            return mpi.mpi_sum_jax(Ol.T @ Or)[0] + self.diag_shift * diag
+        else:
+            # Equivalent to Jr.T@Jr + Ji.T@Ji
+            O = O.reshape(-1, O.shape[-1])
+            return mpi.mpi_sum_jax(O.conj().T @ O)[0] + self.diag_shift * diag
+
+    def to_real_part(self) -> "QGTJacobianDenseT":
+        """
+        Returns the operator computing real part of the complex, non holomorphic QGT.
+
+        The real part of the QGT is used in the Stochastic Reconfiguration (SR)
+        algorithm as well as in the McLachlan or Time-Dependent Variational Principles
+        to simulate Quantum Dynamics.
+
+        .. note::
+
+            This function can only be called on the non-holomorphic QGT.
+
+        .. note::
+
+            The returned object is another :code:`QGTJacobian***` object.
+
+            NetKet does not currently implement the *full* complex QGT object
+            for non-holomorphic functions, instead you need to keep on hand
+            the two separate terms obtained by calling
+            :meth:`~netket.optimizer.qgt.QGTJacobianDenseT.to_real_part` and
+            :meth:`~netket.optimizer.qgt.QGTJacobianDenseT.to_imag_part`.
+
+        See also :meth:`~netket.optimizer.qgt.QGTJacobianDenseT.to_imag_part`.
+        """
+        if self.mode == "imag":
+            return self.replace(mode="complex")
+        elif self.mode == "complex":
+            return self
+        elif self.mode == "real":
+            return self
+        else:
+            raise ValueError(
+                "Can only convert to real part the imaginary part of the"
+                "QGT, not the holomorphic or sign-less one."
+            )
+
+    def to_imag_part(self) -> "QGTJacobianDenseT":
+        """
+        Returns the operator computing imaginary part of the complex, non holomorphic QGT.
+
+        The imaginary part of the QGT is necessary to implement the McLachlan variational
```

Comment:
I am still confused by this, normally for non-holomorphic QGT only the real part is necessary to implement the McLachlan

Reply:
you mean the Dirac TDVP?